---
title: "Protease_2022"
author: "Ella Sieradzki"
date: '2022-07-21'
output: html_document
---
```{r, setup, include=FALSE}
knitr::opts_knit$set(root.dir = '~/Dropbox/PostDoc/MT_nitrogen/')

library("DESeq2")
library("Biobase")
library("ggplot2")
library("data.table")
library("pheatmap")
library("tibble")
library("vegan")
library("gplots")
library("tidyverse")
library("compositions")
library("ape")
library("cowplot")
library("purrr")
library("reshape2")
library("ggpubr")
library("stringr")
library("rstatix")
```


```{r}
# Load data for DESeq2
mtx_counts <- read.csv("Rsubread_output_w_K01430.csv", header = TRUE, row.names = 1)
ann <- read.delim("ann_for_Rsubread_new.txt", sep="\t", header=F, stringsAsFactors = F) %>%
  select(c(1,3))
colnames(ann) <- c("GeneID", "KO")
nitrogen_ko <- c("MER", "K07004", "K07273", "K01428", "K01429", "K01430")
nitrogen_gene <- c("prot","Xds", "lys", "ureA", "ureB", "ureC")
ko_to_gene <- data.frame(nitrogen_ko, nitrogen_gene)
colnames(ko_to_gene) <- c("KO", "gene")

mtx_matrix = as.matrix(sapply(mtx_counts, as.integer)) 
rownames(mtx_matrix) = rownames(mtx_counts)
mtx_matrix[is.na(mtx_matrix)] = 0
mtx_matrix = mtx_matrix[,order(colnames(mtx_matrix))]
colnames(mtx_matrix) <- gsub(".mapped.bam", "", colnames(mtx_matrix))
colnames(mtx_matrix) <- gsub(".", "_", colnames(mtx_matrix), fixed = TRUE)
mtx_matrix <- mtx_matrix[, -(which(colnames(mtx_matrix) %in% c("H3_Rhizo_39", "H1_Rhizo_Litter_2", "H2_Rhizo_Litter_9")))]

designfile = "mtx_design_ecofun_genomes4.txt" 

mtx_design = read.delim(designfile, header = TRUE, row.names = 1)
mtx_design <- mtx_design[!(rownames(mtx_design) %in% c("H3_Rhizo_39", "H1_RhizoLitter_2", "H2_RhizoLitter_9")),] # Remove these because these columns are mostly 0s, was messing up the geometric mean 

colnames(mtx_matrix) <- rownames(mtx_design)
```

```{r}
# Run DESeq2
dds = DESeqDataSetFromMatrix(countData = mtx_matrix, colData = mtx_design, design = ~ microcosm)
dds = dds[ rowSums(counts(dds)) > 5, ] # Filter out rows with < 5 counts

### Create custom groups
dds$group = factor(paste0(dds$treatment, dds$location))
dds$group_harvest = factor(paste0(dds$group, dds$harvest))
dds$group_harvest_location = factor(paste0(dds$harvest, dds$location))
dds$group_harvest_treatment = factor(paste0(dds$harvest, dds$treatment))
dds$group_htl = factor(paste0(dds$harvest, dds$treatment, dds$location))

### Analyze results
design(dds) = ~ group_htl
dds = DESeq(dds)
res = results(dds)
resultsNames(dds)

dds_counts = counts(dds, normalized=TRUE)
rownames(dds_counts) <- rownames(res)
dds_ann <- merge(dds_counts, ann, by.x="row.names", by.y="GeneID")
write.csv(dds_ann, file = "htl_ORFs.csv")
nit <- dds_ann[dds_ann$KO %in% nitrogen_ko,]
nit <- distinct(nit)
write.csv(nit, file = "DESeq_normalized_nitrogen.csv")
saveRDS(dds, file = "dds_htl_nitrogen.rds")
```

```{r}
### Compare to Bulk no litter
res = results(dds)
resultsNames(dds)
res = results(dds, contrast = c("group_htl", "H1LitterRhizo", "H1NoLitterBulk")) # group, numerator, denominator
comparison = "H1NoLitterBulk_vs_H1LitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(res$baseMean)
betas_df = data.frame(res$log2FoldChange)
stat_df = data.frame(res$stat)
padj_df = data.frame(res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat", paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H1LitterBulk", "H1NoLitterBulk"))
comparison = "H1NoLitterBulk_vs_H1LitterBulk"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H1NoLitterRhizo", "H1NoLitterBulk"))
comparison = "H1NoLitterBulk_vs_H1NoLitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H2LitterRhizo", "H2NoLitterBulk"))
comparison = "H2NoLitterBulk_vs_H2LitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H2LitterBulk", "H2NoLitterBulk"))
comparison = "H2NoLitterBulk_vs_H2LitterBulk"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H2NoLitterRhizo", "H2NoLitterBulk"))
comparison = "H2NoLitterBulk_vs_H2NoLitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H3LitterRhizo", "H3NoLitterBulk"))
comparison = "H3NoLitterBulk_vs_H3LitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H3LitterBulk", "H3NoLitterBulk"))
comparison = "H3NoLitterBulk_vs_H3LitterBulk"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H3NoLitterRhizo", "H3NoLitterBulk"))
comparison = "H3NoLitterBulk_vs_H3NoLitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H4LitterRhizo", "H4NoLitterBulk"))
comparison = "H4NoLitterBulk_vs_H4LitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H4LitterBulk", "H4NoLitterBulk"))
comparison = "H4NoLitterBulk_vs_H4LitterBulk"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res = results(dds, contrast = c("group_htl", "H4NoLitterRhizo", "H4NoLitterBulk"))
comparison = "H4NoLitterBulk_vs_H4NoLitterRhizo"
res_sig = which(res$padj < 0.05)
length(res_sig)
baseMean_df = data.frame(baseMean_df, res$baseMean)
betas_df = data.frame(betas_df, res$log2FoldChange)
stat_df = data.frame(stat_df, res$stat)
padj_df = data.frame(padj_df, res$padj)
setnames(baseMean_df, "res.baseMean", paste0("baseMean_", comparison))
setnames(betas_df, "res.log2FoldChange", paste0("betas_", comparison))
setnames(stat_df, "res.stat",paste0("stat_", comparison))
setnames(padj_df, "res.padj", paste0("padj_", comparison))

res_all_data_bulkcomp = data.frame(betas_df, stat_df, padj_df)
rownames(res_all_data_bulkcomp) = rownames(res)
saveRDS(res_all_data_bulkcomp, file = paste(working_directory,"/res_nitrogen_bulkcomp.rds", sep = ""))

# Only keep p values where the comparison by Wald test was significantly different
padj_matrix = as.matrix(res_all_data_bulkcomp[,25:36]) # This is essentially padj_df, but only saved res_all_data, so working from that
padj_matrix[is.na(padj_matrix)] = 1 # Change NA to 1 so will get filtered out in next step
padj_matrix[which(padj_matrix > 0.05)] = NA # Convert all p values > 0.05 to NA
res_all_data_bulkcomp_sig = data.frame(res_all_data_bulkcomp[,1:24], as.data.frame(padj_matrix)) # Remake res_all_data with only significant p values, all non-significant p values are NA

saveRDS(res_all_data_bulkcomp_sig, file = paste(working_directory,"/res_nitrogen_bulkcomp_sig.rds", sep = ""))
#res_litter_LRT_sig_data = readRDS(file = paste(working_directory,"/res_bacteria_bulkcomp_sig.rds", sep = ""))
res_all_data_bulkcomp_sig <- merge(res_all_data_bulkcomp_sig, ann, by.x = "row.names", by.y = "GeneID")
write.csv(res_all_data_bulkcomp_sig[res_all_data_bulkcomp_sig$KO %in% ko_to_gene$KO,], file = paste(working_directory,"results_htl_sig_bulkcomp_nitrogen_080920.csv"), append = FALSE)

```

```{r}
# IDentifying the most highly upregulated proteases
betas <- res_all_data_bulkcomp_sig[, c(1:13, ncol(res_all_data_bulkcomp_sig))]
# Determine protease taxonomy
df <- betas[order(betas$KO),]
write.csv(df, "DESeq_bygene_wKO.csv")

prot_tax <- read.csv("sup_tbl_S5_metaT_vs_prot_reciprocal_besthit.csv")
prot <- df[df$KO=="MER",-c(ncol(df), ncol(df)-1)]
prot <- merge(prot, prot_tax, by.x="Row.names", by.y="transcript_id")
prot_top <- c()
for (c in 3:14) {
  prot <- prot[order(prot[,c], decreasing = T),]
  prot_top <- rbind(prot_top, prot[1:10,])
}
prot_top <- distinct(prot_top)
write.csv(prot_top, "prot_upreg_top.csv")
```

```{r}
# Mapping to curated dataset of genomes - figure 4
MTtoGenomes <- read.csv("database_evan_ulas_sag_genomes.orfs.ko_path_module_hierarchy_cazy2_FIXED_genomic_potential5_Ella.csv", stringsAsFactors = F)
DT <- as.data.table(MTtoGenomes)
MTtoGenomes <- as.data.frame(DT[,which(unlist(lapply(DT, function(x)!all(is.na(x))))),with=F])
MTtoGenomes[, 49:60][is.na(MTtoGenomes[, 49:60])] <- 0
down_cols <- grep("down_sig", colnames(MTtoGenomes))
MTtoGenomes[, down_cols][is.na(MTtoGenomes[, down_cols])] <- 0
MTtoGenomes$down_sig <- rowSums(MTtoGenomes[,down_cols])

prot_IDs <- as.vector(read.delim("prot_metaG.txt", stringsAsFactors = F, header=F))
prot <- MTtoGenomes[MTtoGenomes$id_again %in% prot_IDs$V1 & (MTtoGenomes$up_sig | MTtoGenomes$down_sig),]
prot <- prot[,c(49:60, 2, 75:86, down_cols)]

df <- prot
for (i in 1:12) {
  for (r in 1:nrow(df)) {
    if (is.na(df[r,i+13]) & df[r,i+25]==0) {
      df[r,i] <- 0
    }
  }
}

# Invert betas because Erin runs the analysis as control vs. treatment instead of treatment vs. control
df[,1:12] <- df[,1:12]*-1

# Grouping by bin
df4 <- df %>% group_by(Bin) %>% summarise_each(mean, 1:12)

# To create supplemental figure prot SUM
df5 <- df
df5[1:12] <- 2^df5[1:12]
df5 <- df5 %>% group_by(Bin) %>% summarise_each(max, 1:12)
df5[2:13] <- log2(df5[2:13])

# Throw out random, uninformative bins, change df5 to df4 to recreate figure 6
df5 <- df5[rowSums(df5[,2:13]) >= 1,]
df5 <- df5[-(which(df5$Bin %in% c("SH_S37_Mesorhizobium_64_21","SH_S37_Caulobacter_70_25","Flavobacterium_sp_URHB0058"))),]
sampord <- c(3,6,9,12,2,5,8,11,4,7,10,13)
row.hc <- hclust(dist(as.matrix(df5[,sampord])))
row.dd <- as.dendrogram(row.hc)
weights.dd <-  ifelse(df5$Bin == "YH_S30_unknown_63_8", yes = 50, no = 20) + ifelse(df5$Bin == "b58", yes = 45, no = 20) + ifelse(df5$Bin == "YH_S30_Massilia_66_12", yes = 40, no = 20) + ifelse(df5$Bin == "SM_S39_Burkholderiales_62_29", yes = 35, no = 20) + ifelse(df5$Bin == "b87", yes = 30, no = 20) + ifelse(df5$Bin == "b86", yes = 28, no = 20)
row.dd.reordered <- reorder(row.dd, wts = weights.dd, agglo.FUN = mean)

my_palette <- colorRampPalette(c("blue","lightblue", "white","lightgoldenrod1","red"))(n = 100) # Erin

rem <- c("SH_S37_Mesorhizobium_64_21","YL_S31_Bradyrhizobium_63_8","Bradyrhizobium_URHA0013", "SL_S38_Flavihumibacter_42_5", "SM_S39_Bradyrhizobium_64_9", "b94", "b42", "SH_S37_Burkholderiales_70_21", "Mesorhizobium_URHA0056", "ZL_S28_Bradyrhizobium_63_9", "b75", "SH_S37_Caulobacter_70_25", "Flavobacterium_sp_URHB0058")
df5 <- df4[-(which(df4$Bin %in% rem)),]
row.hc <- hclust(dist(as.matrix(df5[,sampord])))
row.dd <- as.dendrogram(row.hc)
weights.dd <-  ifelse(df5$Bin == "b86", yes = 20, no = 25) + ifelse(df5$Bin == "b87", yes = 20, no = 25) + ifelse(df5$Bin == "b27", yes = 25, no = 15) + ifelse(df5$Bin == "UL_S35_Candidatus_Koribacter_versatilis_56_7", yes = 26, no = 25) + ifelse(df5$Bin == "YH_S30_Massilia_66_12", yes = 30, no = 25) + ifelse(df5$Bin == "YH_S30_Janthinobacterium_sp__HH01_65_15", yes = 28, no = 25) + ifelse(df5$Bin == "SM_S39_Burkholderiales_62_29", yes = 35, no = 20)
row.dd.reordered <- reorder(row.dd, wts = weights.dd, agglo.FUN = mean)

heatmap.2(as.matrix(df5[,sampord]), col=my_palette, cexCol = 0.7, cexRow=0.5, density.info="none", trace="none", labRow = df5$Bin, Rowv = TRUE, Colv=FALSE, dendrogram='row')
```

```{r}
# Protease by taxonomy - figure 3
raw_genes <- read.csv("DESeq_normalized_prot.csv", row.names=1) %>%
  select(-KO) %>%
  column_to_rownames("Row.names")
temp <- t(raw_genes)
prot_bray_dm <- vegdist(temp, method="bray", na.rm=T)
adonis2(prot_bray_dm ~ location * treatment * harvest, data = mtx_design, strata = mtx_design$harvest, permutations = 10000)
tax <- read.csv("sup_tbl_S5_metaT_vs_prot_reciprocal_besthit_EEN.csv", header=T, stringsAsFactors = F) %>%
  select(c(1,2,10,11, 15:22))
raw_genes1 <- merge(raw_genes, tax, by.x="row.names", by.y="transcript_id")
# Get fungi
tmp <- raw_genes1[raw_genes1$Kingdom=="Fungi",]
# Throw out all euks
raw_genes1 <- raw_genes1[!(raw_genes1$Superkingdom %like% "Euk"),]
# Add fungi
raw_genes1 <- rbind.data.frame(raw_genes1, tmp)

# boxplot by taxonomic group - tweak code for other groups
target <- "Fungi"
temp <- raw_genes1 %>% filter(str_detect(Kingdom, target))
temp <- reshape2::melt(temp)
temp <- temp %>% group_by(variable) %>% summarize_at("value", sum)
temp$grp <- sub("_[0-9]*$", "", temp$variable)
temp$time <- sub("_.*$", "", temp$grp)
temp$grp <- sub("H[1-4]_", "", temp$grp)
temp$loc[str_detect(temp$variable, "Bulk")] <- "Bulk"
temp$loc[str_detect(temp$variable, "Rhizo")] <- "Rhizo"
temp$treat[str_detect(temp$variable, "Litter")] <- "Litter"
temp$treat[str_detect(temp$variable, "Litter", negate = T)] <- "NoLitter"
# Box plot
ggplot(data = temp, aes(x = time, y = value)) + 
  ylab("Normalized recruitment to protease") + 
  scale_fill_manual(values=c("#a6cee3", "#1f78b4", "#cab2d6", "#6a3d9a")) + 
  xlab("Time") + 
  geom_boxplot(aes(fill=grp)) + 
  theme_bw() + theme(axis.text.x = element_text(angle = 90), panel.grid=element_blank()) + 
  ggtitle(target) + 
  scale_y_continuous(expand = c(0, 0)) + expand_limits(y = 0) 
ggsave(filename=paste("prot_by_tax/prot_Fungi_n99.pdf", sep=""))
```

```{r}
# Identify the most highly expressed proteases
raw_genes <- raw_genes %>% rownames_to_column("genes")
raw_genes <- reshape2::melt(raw_genes) %>%
  mutate(variable = sub("_[0-9]*$", "", variable))

raw_genes <- as.data.frame(reshape2::dcast(raw_genes, genes~variable, fun.aggregate = mean))
prot <- merge(raw_genes, prot_tax, by.x="genes", by.y="transcript_id")
prot_top <- c()
for (c in 2:17) {
  prot <- prot[order(prot[,c], decreasing = T),]
  prot_top <- rbind(prot_top, prot[1:10,])
}
prot_top <- distinct(prot_top)
write.csv(prot_top, "prot_counts_top.csv")
```

```{r}
# Create sup. fig .1
norm_genes <- read.csv("DESeq_counts_sup_fig_1.csv", row.names=1)
my_palette <- colorRampPalette(c('#ffffff','#8856a7'))(n = 299)
sampord <- c("H1_Bulk", "H2_Bulk", "H3_Bulk", "H4_Bulk", "H1_BulkLitter", "H2_BulkLitter", "H3_BulkLitter", 
             "H4_BulkLitter", "H1_Rhizo", "H2_Rhizo", "H3_Rhizo", "H4_Rhizo", "H1_RhizoLitter", 
             "H2_RhizoLitter", "H3_RhizoLitter", "H4_RhizoLitter")
heatmap.2(as.matrix(norm_genes[,sampord]), col=my_palette, cexCol = 0.8, cexRow=1, density.info="none", trace="none", labRow = norm_genes$gene, Rowv=FALSE, Colv=FALSE, dendrogram = 'none')
```

